intel/geoip/database.go

```diff
@@ -13,106 +13,205 @@ import (
 	"github.com/safing/portmaster/updates"
 )
 
-var (
-	geoDBv4File *updater.File
-	geoDBv6File *updater.File
-	dbFileLock  sync.Mutex
+var worker *updateWorker
 
-	geoDBv4Reader *maxminddb.Reader
-	geoDBv6Reader *maxminddb.Reader
-	dbLock        sync.RWMutex
+func init() {
+	worker = &updateWorker{
+		trigger: make(chan struct{}),
+	}
+}
 
-	dbInUse    = abool.NewBool(false) // only activate if used for first time
-	dbDoReload = abool.NewBool(true)  // if database should be reloaded
+const (
+	v4MMDBResource = "intel/geoip/geoipv4.mmdb.gz"
+	v6MMDBResource = "intel/geoip/geoipv6.mmdb.gz"
 )
 
-// ReloadDatabases reloads the geoip database, if they are in use.
-func ReloadDatabases() error {
-	// don't do anything if the database isn't actually used
-	if !dbInUse.IsSet() {
-		return nil
-	}
+type geoIPDB struct {
+	*maxminddb.Reader
+	file *updater.File
+}
 
-	dbFileLock.Lock()
-	defer dbFileLock.Unlock()
+// updateBroadcaster stores a geoIPDB and provides synchronized
+// access to the MMDB reader. It also supports broadcasting to
+// multiple waiters when a new database becomes available.
+type updateBroadcaster struct {
+	rw sync.RWMutex
+	db *geoIPDB
 
-	dbLock.Lock()
-	defer dbLock.Unlock()
+	waiters []chan struct{}
```

Comment:
Wouldn't a single wait channel do it too? Access to it is locked anyway, so you can just close it and replace it with a new one afterwards.

Reply:
good point :)